lib/inc/internal/substitution_expression.hpp

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include <string>
+#include <memory>
+
+namespace hocon {
+
+    class substitution_expression : public std::enable_shared_from_this<substitution_expression> {
+    public:
+        substitution_expression(std::string path, bool optional);
+
+        std::string path() const;
+        bool optional() const;
+
+        std::shared_ptr<substitution_expression> change_path(std::string new_path);
+
+        std::string to_string() const;
+
+        template <typename T>
```

Comment:
Java uses equals because `==` means exact object equality, and can't be overridden. In C++ if you want to check two objects are the same object you compare their pointers. We also don't need to support general comparison, we shouldn't be trying to compare a `substitution_expression` to a different class.

cpp-hocon has used `operator==` overloading, as in https://github.com/puppetlabs/cpp-hocon/blob/master/lib/inc/hocon/config.hpp#L588 (for limited access) or https://github.com/puppetlabs/cpp-hocon/blob/master/lib/inc/hocon/parser/config_document.hpp#L92 for global access.